File: android/app/src/main/java/com/chatmod/mobile/runtime/CommandRuntime.kt
package com.chatmod.mobile.runtime

import com.chatmod.mobile.youtube.LiveChatMessage
import java.time.Instant

data class BotCommand(
    val id: String,
    val name: String,
    val response: String,
    val aliases: List<String>,
    val cooldownSeconds: Int,
    val accessLevel: CommandAccessLevel,
    val enabled: Boolean
)

enum class CommandAccessLevel {
    Everyone,
    Members,
    Mods,
    Owner
}

data class ParsedCommand(
    val trigger: String,
    val args: List<String>,
    val argText: String
)

data class CommandCooldownState(
    val commandLastUsedAt: Map<String, Long> = emptyMap(),
    val userCommandLastUsedAt: Map<String, Long> = emptyMap()
) {
    fun recordUsage(commandId: String, authorChannelId: String, usedAtMillis: Long): CommandCooldownState {
        val userCommandKey = "$authorChannelId:$commandId"
        return copy(
            commandLastUsedAt = commandLastUsedAt + (commandId to usedAtMillis),
            userCommandLastUsedAt = userCommandLastUsedAt + (userCommandKey to usedAtMillis)
        )
    }
}

data class CommandContext(
    val streamTitle: String = "",
    val streamStartedAt: Instant? = null,
    val now: Instant = Instant.now(),
    val random: () -> Int = { (0..999).random() }
)

data class CommandResult(
    val matched: Boolean,
    val commandId: String?,
    val trigger: String?,
    val response: String?,
    val reason: CommandResultReason,
    val cooldownRemainingSeconds: Int
)

enum class CommandResultReason {
    NotCommand,
    NotFound,
    Disabled,
    AccessDenied,
    Cooldown,
    Matched
}

class CommandRuntime {
    fun parse(text: String): ParsedCommand? {
        val match = Regex("""^!([a-z0-9_-]{1,32})(?:\s+(.*))?$""", RegexOption.IGNORE_CASE)
            .find(text.trim())
            ?: return null
        val argText = match.groupValues.getOrNull(2)?.trim().orEmpty()

        return ParsedCommand(
            trigger = "!${match.groupValues[1].lowercase()}",
            args = if (argText.isBlank()) emptyList() else argText.split(Regex("""\s+""")),
            argText = argText
        )
    }

    fun evaluate(
        message: LiveChatMessage,
        commands: List<BotCommand>,
        cooldownState: CommandCooldownState = CommandCooldownState(),
        context: CommandContext = CommandContext()
    ): CommandResult {
        val parsed = parse(message.text) ?: return noMatch(CommandResultReason.NotCommand)
        val command = commands.firstOrNull { it.matches(parsed.trigger) }
            ?: return noMatch(CommandResultReason.NotFound, parsed.trigger)

        if (!command.enabled) {
            return CommandResult(false, command.id, parsed.trigger, null, CommandResultReason.Disabled, 0)
        }

        if (!message.canUse(command.accessLevel)) {
            return CommandResult(false, command.id, parsed.trigger, null, CommandResultReason.AccessDenied, 0)
        }

        val cooldownRemaining = cooldownRemainingSeconds(message, command, cooldownState, context.now)
        if (cooldownRemaining > 0) {
            return CommandResult(false, command.id, parsed.trigger, null, CommandResultReason.Cooldown, cooldownRemaining)
        }

        return CommandResult(
            matched = true,
            commandId = command.id,
            trigger = parsed.trigger,
            response = render(command.response, message, parsed, context),
            reason = CommandResultReason.Matched,
            cooldownRemainingSeconds = 0
        )
    }

    private fun noMatch(reason: CommandResultReason, trigger: String? = null): CommandResult {
        return CommandResult(
            matched = false,
            commandId = null,
            trigger = trigger,
            response = null,
            reason = reason,
            cooldownRemainingSeconds = 0
        )
    }

    private fun BotCommand.matches(trigger: String): Boolean {
        return name.lowercase() == trigger || aliases.any { it.lowercase() == trigger }
    }

    private fun LiveChatMessage.canUse(accessLevel: CommandAccessLevel): Boolean {
        return when (accessLevel) {
            CommandAccessLevel.Everyone -> true
            CommandAccessLevel.Members -> isOwner || isModerator || isMember
            CommandAccessLevel.Mods -> isOwner || isModerator
            CommandAccessLevel.Owner -> isOwner
        }
    }

    private fun cooldownRemainingSeconds(
        message: LiveChatMessage,
        command: BotCommand,
        cooldownState: CommandCooldownState,
        now: Instant
    ): Int {
        if (command.cooldownSeconds <= 0) return 0

        val userCommandKey = "${message.authorChannelId}:${command.id}"
        val lastUsedAt = listOfNotNull(
            cooldownState.commandLastUsedAt[command.id],
            cooldownState.userCommandLastUsedAt[userCommandKey]
        ).maxOrNull() ?: return 0

        val elapsedSeconds = ((now.toEpochMilli() - lastUsedAt) / 1000L).toInt()
        return (command.cooldownSeconds - elapsedSeconds).coerceAtLeast(0)
    }

    private fun render(
        template: String,
        message: LiveChatMessage,
        parsed: ParsedCommand,
        context: CommandContext
    ): String {
        return template
            .replace("{username}", message.authorName)
            .replace("{args}", parsed.argText)
            .replace("{streamTitle}", context.streamTitle)
            .replace("{uptime}", formatUptime(context.streamStartedAt, context.now))
            .replace("{time}", context.now.toString())
            .replace("{random}", context.random().toString())
    }

    private fun formatUptime(streamStartedAt: Instant?, now: Instant): String {
        if (streamStartedAt == null) {
            return "not available yet"
        }

        val totalMinutes = ((now.toEpochMilli() - streamStartedAt.toEpochMilli()) / 60_000L).coerceAtLeast(0L)
        val hours = totalMinutes / 60L
        val minutes = totalMinutes % 60L
        return if (hours > 0) {
            "${hours}h ${minutes}m"
        } else {
            "${minutes}m"
        }
    }
}

File: backend/src/modules/commands/commandRuntime.ts
import type { CommandRecord } from "./commandStore.js";

export interface ChatCommandMessage {
  authorChannelId: string;
  authorName: string;
  text: string;
  isOwner?: boolean;
  isModerator?: boolean;
  isMember?: boolean;
}

export interface CommandRuntimeContext {
  streamTitle?: string;
  streamStartedAt?: Date;
  now?: Date;
  random?: () => number;
}

export interface CommandCooldownState {
  commandLastUsedAt?: Record<string, string>;
  userCommandLastUsedAt?: Record<string, string>;
}

export interface ParsedCommand {
  trigger: string;
  args: string[];
  argText: string;
}

export interface CommandEvaluation {
  matched: boolean;
  commandId: string | null;
  trigger: string | null;
  response: string | null;
  reason: "not_command" | "not_found" | "disabled" | "access_denied" | "cooldown" | "matched";
  cooldownRemainingSeconds: number;
}

const commandPattern = /^!([a-z0-9_-]{1,32})(?:\s+(.*))?$/i;

export function parseCommandMessage(text: string): ParsedCommand | null {
  const match = text.trim().match(commandPattern);
  if (!match) {
    return null;
  }

  const argText = match[2]?.trim() ?? "";

  return {
    trigger: `!${match[1].toLowerCase()}`,
    args: argText.length > 0 ? argText.split(/\s+/g) : [],
    argText
  };
}

export function evaluateCommand(
  message: ChatCommandMessage,
  commands: CommandRecord[],
  cooldownState: CommandCooldownState = {},
  context: CommandRuntimeContext = {}
): CommandEvaluation {
  const parsed = parseCommandMessage(message.text);
  if (!parsed) {
    return noMatch("not_command");
  }

  const command = commands.find((candidate) => commandMatches(candidate, parsed.trigger));
  if (!command) {
    return noMatch("not_found", parsed.trigger);
  }

  if (!command.enabled) {
    return {
      matched: false,
      commandId: command.id,
      trigger: parsed.trigger,
      response: null,
      reason: "disabled",
      cooldownRemainingSeconds: 0
    };
  }

  if (!canUseCommand(message, command.accessLevel)) {
    return {
      matched: false,
      commandId: command.id,
      trigger: parsed.trigger,
      response: null,
      reason: "access_denied",
      cooldownRemainingSeconds: 0
    };
  }

  const cooldownRemainingSeconds = remainingCooldownSeconds(message, command, cooldownState, context.now ?? new Date());
  if (cooldownRemainingSeconds > 0) {
    return {
      matched: false,
      commandId: command.id,
      trigger: parsed.trigger,
      response: null,
      reason: "cooldown",
      cooldownRemainingSeconds
    };
  }

  return {
    matched: true,
    commandId: command.id,
    trigger: parsed.trigger,
    response: renderCommandResponse(command.response, message, parsed, context),
    reason: "matched",
    cooldownRemainingSeconds: 0
  };
}

export function cooldownKeys(message: ChatCommandMessage, command: CommandRecord): {
  commandKey: string;
  userCommandKey: string;
} {
  return {
    commandKey: command.id,
    userCommandKey: `${message.authorChannelId}:${command.id}`
  };
}

function noMatch(reason: CommandEvaluation["reason"], trigger: string | null = null): CommandEvaluation {
  return {
    matched: false,
    commandId: null,
    trigger,
    response: null,
    reason,
    cooldownRemainingSeconds: 0
  };
}

function commandMatches(command: CommandRecord, trigger: string): boolean {
  return command.name.toLowerCase() === trigger || command.aliases.some((alias) => alias.toLowerCase() === trigger);
}

function canUseCommand(message: ChatCommandMessage, accessLevel: CommandRecord["accessLevel"]): boolean {
  if (accessLevel === "everyone") {
    return true;
  }
  if (accessLevel === "members") {
    return Boolean(message.isMember || message.isModerator || message.isOwner);
  }
  if (accessLevel === "mods") {
    return Boolean(message.isModerator || message.isOwner);
  }
  if (accessLevel === "owner") {
    return Boolean(message.isOwner);
  }

  return false;
}

function remainingCooldownSeconds(
  message: ChatCommandMessage,
  command: CommandRecord,
  cooldownState: CommandCooldownState,
  now: Date
): number {
  if (command.cooldownSeconds <= 0) {
    return 0;
  }

  const keys = cooldownKeys(message, command);
  const lastUsedAt = [
    cooldownState.commandLastUsedAt?.[keys.commandKey],
    cooldownState.userCommandLastUsedAt?.[keys.userCommandKey]
  ]
    .filter(Boolean)
    .map((value) => new Date(value!))
    .filter((value) => !Number.isNaN(value.getTime()))
    .sort((a, b) => b.getTime() - a.getTime())[0];

  if (!lastUsedAt) {
    return 0;
  }

  const elapsedSeconds = Math.floor((now.getTime() - lastUsedAt.getTime()) / 1000);
  return Math.max(0, command.cooldownSeconds - elapsedSeconds);
}

function renderCommandResponse(
  template: string,
  message: ChatCommandMessage,
  parsed: ParsedCommand,
  context: CommandRuntimeContext
): string {
  const now = context.now ?? new Date();
  const random = Math.floor((context.random?.() ?? Math.random()) * 1000).toString();

  return template
    .replaceAll("{username}", message.authorName)
    .replaceAll("{args}", parsed.argText)
    .replaceAll("{streamTitle}", context.streamTitle ?? "")
    .replaceAll("{uptime}", formatUptime(context.streamStartedAt, now))
    .replaceAll("{time}", now.toISOString())
    .replaceAll("{random}", random);
}

function formatUptime(streamStartedAt: Date | undefined, now: Date): string {
  if (!streamStartedAt || Number.isNaN(streamStartedAt.getTime())) {
    return "not available yet";
  }

  const totalMinutes = Math.max(0, Math.floor((now.getTime() - streamStartedAt.getTime()) / 60000));
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  if (hours > 0) {
    return `${hours}h ${minutes}m`;
  }
  return `${minutes}m`;
}
